### PR TITLE
editorial: Introduce clocks, moments, and durations

### LINKED ISSUE #139: Spec text isn't clear about time origins and monotonic clocks
In spec order:

* "The [time origin](https://w3c.github.io/hr-time/#dfn-time-origin) is the time value from which time is measured" <- what is a time value?
* [get time origin timestamp](https://w3c.github.io/hr-time/#dfn-get-time-origin-timestamp): "Let t1 be the DOMHighResTimeStamp representing the high resolution time at which the shared monotonic clock is zero." <- what is a high resolution time? The fact that it's a DOMHighResTimeStamp doesn't help, since those have 3 possible meanings: "measured relative from the [time origin](https://w3c.github.io/hr-time/#dfn-time-origin), [shared monotonic clock](https://w3c.github.io/hr-time/#dfn-shared-monotonic-clock), or a time value".
* ["the high resolution time value at which time origin is zero."](https://w3c.github.io/hr-time/#issue-container-generatedID-1) <- It doesn't make sense to talk about a time at which "the time value from which time is measured" is zero.
* [Monotonic clock](https://w3c.github.io/hr-time/#dfn-monotonic-clock) and [shared monotonic clock](https://w3c.github.io/hr-time/#dfn-shared-monotonic-clock) aren't really defined. 
* The sentences "monotonic clock" and "shared monotonic clock" appear in are COMEFROM requirements: in order to have an effect, they need to be placed inside the definitions of [now](https://w3c.github.io/hr-time/#dom-performance-now)() and [timeOrigin](https://w3c.github.io/hr-time/#dom-performance-timeorigin), respectively.
* It's not clear what "not subject to ... system clock skew" would mean. If it's just that the monotonic clock's values aren't comparable across machines, we should probably say that directly. 
* "whose reference point is the [ECMA-262] [time](https://tc39.es/ecma262/multipage/#sec-time-values-and-time-range) definition" is the only use of "reference point" in this document or ECMA-262. What do we mean by it? If it's that the shared monotonic clock needs to measure time with its 0 at the epoch, and needs to jump at leap seconds, the latter conflicts with the idea of a monotonic clock.

### PR DESCRIPTION
**Pull request by jyasskin**
I believe this fixes #139.

***
<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html" title="Last updated on Jan 12, 2023, 3:15 PM UTC (2c194b4)">Preview</a> (<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#sec-concepts" title="#sec-concepts">#sec-…</a>) (<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#sec-tools" title="#sec-tools">#sec-…</a>) (<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#sec-time-origin" title="#sec-time-origin">#sec-…</a>) (<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#sec-domhighrestimestamp" title="#sec-domhighrestimestamp">#sec-…</a>) (<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#the-epochtimestamp-typedef" title="#the-epochtimestamp-typedef">#the-…</a>) (<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#sec-performance" title="#sec-performance">#sec-…</a>) (<a href="https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#clock-drift" title="#clock-drift">#cloc…</a>) | <a href="https://pr-preview.s3.amazonaws.com/w3c/hr-time/140/f1fd2aa...jyasskin:2c194b4.html" title="Last updated on Jan 12, 2023, 3:15 PM UTC (2c194b4)">Diff</a>

### CODE REVIEW

**Comment by noamr** [comment]
<bikeshedding> I find the term "instant" unfamiliar and confusing as it is more widely used as an adjective. Anything wrong with `timestamp`, or perhaps `moment`?

File: index.html [hunk 1]
```diff
@@ -231,17 +231,112 @@ <h3>
...
+          clock=] sometimes needs to be adjusted, which means the [=wall
+          clock/current time=] can decrease, making it unreliable for performance
+          measurement or recording the orders of events. [[ECMA-262]] <dfn
+          data-no-export="" data-cite=
+          "ECMA-262#sec-time-values-and-time-range">time</dfn> is a [=wall
+          clock=].</li>
+
+          <li>A <dfn data-export>monotonic clock</dfn>'s current time never
+          decreases, so it can't be changed by system clock adjustments. There are
+          several kinds of [=monotonic clock=] varying by the point in time they
```
Comment (yoavweiss) [inline index.html:264]: Are there multiple web exposed monotonic clocks? If not, we probably shouldn't mention them.

File: index.html [hunk 2]
```diff
@@ -231,17 +231,112 @@ <h3>
...
+          assign to 0, whether they pause while the computer is asleep, and whether
+          their speed adjusts to keep them close to the [=wall clock=].</li>
+
+          <li>
+            <p>
+              The <dfn>shared monotonic clock</dfn> is a [=monotonic clock=]
+              that's shared between all [=browsing contexts=] that can communicate
+              with each other. Its <dfn data-export data-dfn-for="shared monotonic
```
Comment (yoavweiss) [inline index.html:272]: Should we use something like [familiar with](https://html.spec.whatwg.org/#familiar-with) to make the "can communicate with each other" part a bit more well defined?

Comment (yoavweiss) [reply index.html:272]: "familiar with" is probably not the right call, as workers are also part of this. But we should probably find a concept that matches what we need here.

File: index.html [hunk 3]
```diff
@@ -231,17 +231,112 @@ <h3>
...
+              clock">current time</dfn> will advance monotonically and be comparable
+              across all of those browsing contexts.
+            </p>
+            <p class="note">
+              The user agent can reset its <a>shared monotonic clock</a> across
+              browser restarts, or whenever starting an isolated browsing
+              session—e.g. incognito or similar browsing mode. As a result,
```
Comment (yoavweiss) [inline index.html:279]: The UA can also reset the shared monotonic clock when navigating to a new browsing context that has no other context they can communicate with, right?

File: index.html [hunk 4]
```diff
@@ -256,33 +351,28 @@ <h3>
...
       <p class="note">
-        The value returned by [=get time origin timestamp=] is the high
-        resolution time value at which <a>time origin</a> is zero. It may
-        differ from the value returned by <a>Date.now()</a> executed at "zero
-        time", because the former is recorded with respect to a <a>shared
-        monotonic clock</a> that is not subject to system and user clock
-        adjustments, clock skew, and so on — see <a href=
-        "#sec-monotonic-clock"></a>.
+        The value returned by [=get time origin timestamp=] is the [=wall
+        clock=] time value at |global|'s [=environment settings object/time
```
Comment (yoavweiss) [inline index.html:367]: I don't think we want the time origin to be based on the wall clock.

File: index.html (hunk 3 above)
Comment (yoavweiss) [reply index.html:279]: Should we define how the shared monotonic clock gets reset?

File: index.html (hunk 4 above)
Comment (jyasskin) [reply index.html:367]: I think the current spec expects `performance.timeOrigin` to approximate a wall clock time. "shared monotonic clock ... whose reference point is the [[ECMA-262](https://w3c.github.io/hr-time/#bib-ecma-262)] [time](https://tc39.es/ecma262/multipage/#sec-time-values-and-time-range) definition" probably means the epochs must match, and the informative text about "the difference between `Date.now() - performance.now()` and `performance.timeOrigin`" and «the value returned by [Date.now()](https://tc39.es/ecma262/multipage/#sec-date.now) executed at "zero time"» also imply that `.timeOrigin` comes from a `Date.now()` call. Firefox, Chrome, and Safari also match that behavior.

**Review by jyasskin — COMMENTED** [review]
Ok, what do you think, especially about the new tools for specification authors?

**Comment by jyasskin** [comment]
What's the next step for this PR? Should I present it to the WebPerfWG or its mailing list, or are the current reviewers enough? @nicjansma @yoavweiss?

**Comment by jyasskin** [comment]
What's the best way to push this forward? @yoavweiss mentioned wanting @noamr to review this...

**Comment by csharrison** [comment]
Just chiming in to say that we would like to use these spec concepts (especially the wall clock and unix time) in the [attribution reporting api spec](https://wicg.github.io/attribution-reporting-api/). Right now we just say "seconds since the unix epoch" in a few places, which isn't great.

**Comment by noamr** [comment]
> What's the best way to push this forward? @yoavweiss mentioned wanting @noamr to review this...

Sorry, this slipped through the xmas cracks. Will review this week.

File: index.html [hunk 5]
```diff
@@ -231,17 +239,269 @@ <h3>
...
+      </p>
+      <p>
+        [=Moments=] from the [=monotonic clock=] can't be directly represented
+        in Javascript or HTTP. Instead, expose a [=duration=] between two such
+        [=moments=].
+      </p>
+      <p>
+        For measuring time across multiple UA executions, create [=moments=]
+        using a |settingsObject|'s <dfn data-export
+        data-dfn-for="environment settings object">current wall time</dfn>,
```
Comment (noamr) [inline index.html:362]: Instead of "defined as" and "applied to", perhaps something more direct: the steps to compute the current wall time are to coarsen...

**Review by noamr — COMMENTED** [review]
Great writeup, added some comments. We're pretty close.

**Review by noamr — APPROVED** [review]
LGTM % nit
@yoavweiss is this good with you? Want another round?
Asking because it's a lengthy patch...

File: index.html [hunk 6]
```diff
@@ -231,17 +239,293 @@ <h3>
...
+        <p>
+          Each [=clock=]'s [=unsafe current time=] returns an <dfn>unsafe
+          moment</dfn>. [=Coarsen time=] converts these [=unsafe moments=] to
+          <dfn data-export data-lt="moment">coarsened moments</dfn> or just
+          [=moments=]. [=Unsafe moments=] and [=moments=] from different
+          clocks are not comparable.
+        </p>
+        <p>
+          A <dfn data-export>duration</dfn> is the distance from one [=moment=]
+          to another [=moment=] from the same [=clock=]. One of the two
```
[the remark this answers is not in the document]
Comment (jyasskin) [reply index.html:317]: Hm, this reads clearly to me, which means I might have completely failed to communicate the idea. 😅   I wanted to make a rule that *durations* are always safe, by requiring that at least one of their endpoints has been coarsened. I can try to reword this to say that directly tomorrow, or I'm happy to take suggested wording.

[the remark this answers is not in the document]
Comment (noamr) [reply index.html:317]: Ah that totally got lost on me and is an important bit.
I think we need to split `duration` into two:
- A definition of what it is
- An algorithm that takes two moments, asserts that at least one of them is safe, and returns a - b or so

File: index.html [hunk 7]
```diff
@@ -231,17 +239,293 @@ <h3>
...
+        <li>Let |monotonic time| be the [=monotonic clock=]'s [=monotonic
+        clock/unsafe current time=].
+        </li>
+        <li>Initialize the [=estimated monotonic time of the Unix epoch=] to
+        <code>|monotonic time| - (|wall time| - [=Unix epoch=])</code>.
+        </li>
+      </ol>
+      <div class="issue">
+        The above set of settings-objects-that-could-possibly-communicate needs
+        to be specified better. It's similar to <a
```
Comment (yoavweiss) [inline index.html:516]: +1

**Review by yoavweiss — COMMENTED** [review]
A bunch of comments and questions, but all in all looks great!!

File: index.html [hunk 8]
```diff
@@ -231,17 +239,293 @@ <h3>
...
+        data-export data-dfn-for="environment settings object">current relative
+        timestamp</dfn>, defined as the [=duration=] from |settingsObject|'s
+        [=environment settings object/time origin=] to the |settingsObject|'s
+        [=environment settings object/current monotonic time=]. This value can
+        be exposed directly to JavaScript using the [=duration=]'s [=implicitly
+        convert a duration to a timestamp|implicit conversion=] to
+        {{DOMHighResTimeStamp}}.
+      </p>
+      <p>
+        For measuring time within a single UA execution when an [=environment
```
[the remark this answers is not in the document]
Comment (jyasskin) [reply index.html:362]: https://www.w3.org/TR/reporting-1/#serialize-a-list-of-reports-to-json does this. I think it would be possible in that case to compare the "current relative timestamp" at report generation vs serialization, but that's harder to think about since you wind up keeping track of 3 points in time rather than 2.

File: index.html [hunk 9]
```diff
@@ -231,17 +239,293 @@ <h3>
...
+        data-dfn-for="environment settings object">current wall time</dfn> is
+        the result of the following steps:
+      </p>
+      <ol>
+        <li>
+          Let |unsafeWallTime| be the [=wall clock=]'s [=wall clock/unsafe
+          current time=].
+        </li>
+        <li>
+          Return the result of calling [=coarsen time=] with |unsafeWallTime|
```
[the remark this answers is not in the document]
Comment (jyasskin) [reply index.html:400]: `Date.now()` is defined by https://tc39.es/ecma262/#sec-date.now, so we'd need to move all this text back to Infra or over to ECMAScript in order to use it in that definition. I think we'll be fine saying they're meant to be compatible (as I do in https://pr-preview.s3.amazonaws.com/jyasskin/hr-time/pull/140.html#dfn-wall-clock) but letting their definitions stay separate. If TC39 wants to merge later, we can refactor then.

We're likely to add call sites for this algorithm everywhere [https://wicg.github.io/attribution-reporting-api/](https://wicg.github.io/attribution-reporting-api/#:~:text=current%20time) uses "current time".

File: index.html [hunk 10]
```diff
@@ -231,17 +239,293 @@ <h3>
...
+            </p>
+            <p class="note">
+              The user agent can pick a new [=estimated monotonic time of the
+              Unix epoch=] when the browser restarts, when it starts an isolated
+              browsing session—e.g. incognito or a similar browsing mode—or when
+              it creates an [=environment settings object=] that can't
+              communicate with any existing settings objects. As a result,
+              developers should not use shared timestamps as absolute time that holds
+              its monotonic properties across all past, present, and future contexts;
+              in practice, the monotonic properties only apply for contexts that can
```
[the remark this answers is not in the document]
Comment (jyasskin) [reply index.html:290]: I see you already have an issue for this: https://github.com/w3c/hr-time/issues/141

File: index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (yoavweiss) [reply index.html]: LG!

File: index.html [hunk 11]
```diff
@@ -231,17 +239,293 @@ <h3>
...
+
+          <li>
+            <p>
+              The <dfn data-export>monotonic clock</dfn>'s <dfn data-export
+              data-dfn-for="monotonic clock">unsafe current time</dfn> never decreases,
+              so it can't be changed by system clock adjustments. It only exists
+              within a single execution of the [=user agent=], so it can't be
+              used to compare events that might happen in different executions.
+              Since it can't be adjusted to match the user's notion of time,
+              it's also not appropriate for communicating with the user.
```
[the remark this answers is not in the document]
Comment (yoavweiss) [reply index.html]: Maybe "The monotonic clock should be used for measurement, rather than user-visible timers. For any time communication with the user, the wall clock should be used" as a separate note? (as this part is essentially directed at users of the APIs rather than implementers)

File: index.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (yoavweiss) [reply index.html:362]: OK, cool

File: index.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (yoavweiss) [reply index.html:400]: SG!

**Review by yoavweiss — APPROVED** [review]
LGTM % note for monotonic clock users

File: index.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (jyasskin) [reply index.html]: How's this?

[the remark this answers is not in the document]
Comment (yoavweiss) [reply index.html]: Looks great!!

[the remark this answers is not in the document]
Comment (jyasskin) [reply index.html]: In that case, please merge at your earliest convenience. 🙂
